File: src/test/java/com/github/urlshorten/junit4/UrlShortenServiceTest.java
package com.github.urlshorten.junit4;

import com.github.urlshorten.Url;
import com.github.urlshorten.UrlRepository;
import com.github.urlshorten.UrlShortenService;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import java.util.Optional;

import static org.junit.Assert.assertSame;
import static org.junit.Assert.assertTrue;
import static org.mockito.BDDMockito.*;

public class UrlShortenServiceTest {

  @Mock
  UrlRepository repository;

  UrlShortenService service;

  @Before
  public void set_up() {
    MockitoAnnotations.initMocks(this);

    service = new UrlShortenService(repository);
  }

  @After
  public void tear_down() {
    verifyNoMoreInteractions(repository);
  }

  @Test
  public void should_retrieve() {
    // given
    var url = "https://google.com";
    var entity = new Url();
    given(repository.findByShortenUrl(anyString())).willReturn(Optional.of(entity));

    // when
    var optUrl = service.retrieve(url);

    // then
    assertTrue(optUrl.isPresent());
    assertSame(entity, optUrl.get());

    verify(repository).findByShortenUrl(url);
  }
}

File: src/test/kotlin/com/github/urlshorten/junit5/UrlShortenServiceTest.kt
package com.github.urlshorten.junit4

import com.github.urlshorten.Url
import com.github.urlshorten.UrlRepository
import com.github.urlshorten.UrlShortenService
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.*
import java.time.LocalDateTime

class UrlShortenServiceTest {
    @Mock
    lateinit var repository: UrlRepository
    lateinit var service: UrlShortenService

    @BeforeEach
    fun set_up() {
        MockitoAnnotations.initMocks(this)
        service = UrlShortenService(repository)
    }

    @AfterEach
    fun tear_down() {
        Mockito.verifyNoMoreInteractions(repository)
    }

    @Test
    fun should_retrieve() {
        // given
        val url = "https://google.com"
        val entity = Url(
                shortenUrl = null,
                originalUrl = "",
                expirationDate = LocalDateTime.now(),
                creationDate = LocalDateTime.now(),
                id = null,
        )
        BDDMockito.given(repository.findByShortenUrl(ArgumentMatchers.anyString())).willReturn(entity)

        // when
        val optUrl = service.retrieve(url)

        // then
        Assertions.assertSame(entity, optUrl)
        Mockito.verify(repository)!!.findByShortenUrl(url)
    }
}